File: app/src/main/java/com/team2052/frckrawler/fragments/event/EventInfoFragment.java
package com.team2052.frckrawler.fragments.event;

import android.os.AsyncTask;
import android.os.Bundle;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.AppCompatEditText;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.team2052.frckrawler.R;
import com.team2052.frckrawler.activities.BaseActivity;
import com.team2052.frckrawler.background.DeleteEventTask;
import com.team2052.frckrawler.db.Event;
import com.team2052.frckrawler.fragments.BaseFragment;
import com.team2052.frckrawler.listeners.ListUpdateListener;
import com.team2052.frckrawler.util.Util;

/**
 * Created by adam on 6/15/15.
 */
public class EventInfoFragment extends BaseFragment implements ListUpdateListener {
    public static final String EVENT_ID = "EVENT_ID";
    private Event mEvent;

    public static EventInfoFragment newInstance(Event event) {
        Bundle args = new Bundle();
        args.putLong(EVENT_ID, event.getId());
        EventInfoFragment fragment = new EventInfoFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
        mEvent = mDbManager.getEventsTable().load(getArguments().getLong(EVENT_ID));
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_event_info, null);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        //binding = FragmentEventInfoBinding.bind(view);
        updateList();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.edit_delete_menu, menu);
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_delete:
                buildDeleteDialog().show();
                break;
            case R.id.menu_edit:
                buildEditDialog().show();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    public AlertDialog buildDeleteDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        builder.setTitle("Delete Event?");
        builder.setMessage("Are you sure you want to delete this event?");
        builder.setPositiveButton("Ok", (dialog, which) -> {
            new DeleteEventTask(getActivity(), mEvent, true).execute();
        });
        builder.setNegativeButton("Cancel", null);
        return builder.create();
    }

    public AlertDialog buildEditDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        AppCompatEditText name = new AppCompatEditText(getActivity());
        name.setText(mEvent.getName());
        int padding = Util.getPixelsFromDp(getActivity(), 16);
        name.setPadding(padding, padding, padding, padding);
        builder.setView(name);
        builder.setTitle("Edit Event");
        builder.setPositiveButton("Ok", (dialog, which) -> {
            mEvent.setName(name.getText().toString());
            mEvent.update();
            ((BaseActivity) getActivity()).setActionBarSubtitle(mEvent.getName());
        });
        builder.setNegativeButton("Cancel", null);
        return builder.create();
    }

    @Override
    public void updateList() {
        new LoadEventInfo().execute();
    }

    public class LoadEventInfo extends AsyncTask<Void, Void, Void> {
        int numOfTeams = 0;
        int numOfMatches = 0;
        int numOfPitData = 0;
        int numOfMatchData = 0;

        @Override
        protected Void doInBackground(Void... params) {
            numOfTeams = mDbManager.getEventsTable().getRobotEvents(mEvent).size();
            numOfMatches = mDbManager.getEventsTable().getMatches(mEvent).size();
            numOfPitData = mDbManager.getEventsTable().getPitData(mEvent).size();
            numOfMatchData = mDbManager.getEventsTable().getMatchData(mEvent).size();
            return null;
        }

        @Override
        protected void onPostExecute(Void aVoid) {
            /*binding.setNumOfTeams(numOfTeams);
            binding.setNumOfMatches(numOfMatches);
            binding.setNumOfPitData(numOfPitData);
            binding.setNumOfMatchData(numOfMatchData);*/
        }
    }
}

File: app/src/main/java/com/team2052/frckrawler/fragments/event/EventInfoFragment.kt
package com.team2052.frckrawler.fragments.event

import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.widget.AppCompatEditText
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem

import com.team2052.frckrawler.R
import com.team2052.frckrawler.activities.NavigationDrawerActivity
import com.team2052.frckrawler.di.subscribers.KeyValueListSubscriber
import com.team2052.frckrawler.fragments.ListViewFragment
import com.team2052.frckrawler.helpers.Util
import com.team2052.frckrawler.interfaces.RefreshListener
import com.team2052.frckrawler.models.Event

import rx.Observable
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers

class EventInfoFragment : ListViewFragment<Map<String, String>, KeyValueListSubscriber>(), RefreshListener {
    private var mEvent: Event? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
        mEvent = rxDbManager.eventsTable.load(arguments?.getLong(EVENT_ID))
    }

    override fun inject() {
        mComponent.inject(this)
    }

    override fun getObservable(): Observable<out Map<String, String>> {
        return rxDbManager.eventInfo(mEvent)
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        inflater!!.inflate(R.menu.edit_delete_menu, menu)
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item!!.itemId) {
            R.id.menu_delete -> buildDeleteDialog()?.show()
            R.id.menu_edit -> buildEditDialog()?.show()
        }
        return super.onOptionsItemSelected(item)
    }

    private fun buildDeleteDialog(): AlertDialog? {
        val act = activity ?: return null

        val builder = AlertDialog.Builder(act).apply {
            setTitle("Delete Event?")
            setMessage("Are you sure you want to delete this event?")
            setPositiveButton("Ok") { dialog, which ->
                Observable.just<Event>(mEvent)
                        .map { event ->
                            rxDbManager.runInTx(Runnable { rxDbManager.eventsTable.delete(event) })
                            event
                        }
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeOn(Schedulers.io())
                        .subscribe { activity?.finish() }
            }
            setNegativeButton("Cancel", null)
        }
        return builder.create()
    }

    private fun buildEditDialog(): AlertDialog? {
        val act = activity ?: return null

        val builder = AlertDialog.Builder(act).apply {
            val name = AppCompatEditText(activity)
            name.setText(mEvent!!.name)
            val padding = Util.getPixelsFromDp(activity, 16)
            name.setPadding(padding, padding, padding, padding)
            setView(name)
            setTitle("Edit Event")
            setPositiveButton("Ok") { _, _ ->
                mEvent?.let {
                    it.name = name.text.toString()
                    it.update()
                }
                (activity as NavigationDrawerActivity).setActionBarSubtitle(mEvent?.name)
            }
            setNegativeButton("Cancel", null)
        }
        return builder.create()
    }

    companion object {
        val EVENT_ID = "EVENT_ID"

        fun newInstance(event_id: Long): EventInfoFragment {
            val args = Bundle()
            args.putLong(EVENT_ID, event_id)
            val fragment = EventInfoFragment()
            fragment.arguments = args
            return fragment
        }
    }
}
